<a href="https://colab.research.google.com/github/dhdbsrlw/kupply-MLOps/blob/main/dataPreprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **모의지원 합격 안정도 예측 모델 데이터 전처리 - 4 CLASS**

Multi-class Classification Task - *unsupervised learning*

[ 특이사항 ]
*   데이터에 한국어 및 특수기호(-) 포함
*   설문조사 데이터에는 라벨이 있지만, skewed data (합격 샘플 多)




## 1. 데이터 전처리

**대상 데이터**: 1) 회원가입 데이터  2) 모의지원 데이터 \
**모델**: Clustering (ML 모델)

\
+) 나중에 설문조사 데이터를 바탕으로 CLS 모델도 만들어볼 수 있다. \
++) labelled data 와 unlabelled data 를 혼합하여 CLS 모델을 만들어볼 수 있다.

In [ ]:
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
import json

### 1.1 JSON 파일 데이터프레임으로 저장

In [ ]:
# 회원가입 정보 JSON 읽어오기

with open('/content/drive/MyDrive/Colab Notebooks/kuply-MLOps/users.json') as f:
    dataset = json.loads(f.read())
    dataset = dataset['users']

print(dataset)
# user_df = pd.DataFrame(data)


In [ ]:
# 데이프레임화

user_df = pd.DataFrame(dataset)
user_df.head()

In [ ]:
user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 242 entries, 0 to 241
Data columns (total 21 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   _id           242 non-null    object 
 1   name          242 non-null    object 
 2   studentId     242 non-null    object 
 3   email         242 non-null    object 
 4   firstMajor    242 non-null    object 
 5   nickname      242 non-null    object 
 6   role          242 non-null    object 
 7   totalReport   242 non-null    int64  
 8   profilePic    242 non-null    object 
 9   leave         242 non-null    bool   
 10  secondMajor   19 non-null     object 
 11  passSemester  19 non-null     object 
 12  passGPA       19 non-null     float64
 13  __v           242 non-null    int64  
 14  hopeMajor1    223 non-null    object 
 15  hopeMajor2    223 non-null    object 
 16  hopeSemester  223 non-null    object 
 17  curGPA        223 non-null    float64
 18  changeGPA     223 non-null    

In [ ]:
# 모의지원 정보 JSON 읽어오기
with open('/content/drive/MyDrive/Colab Notebooks/kuply-MLOps/applicaions.json') as f:
    dataset = json.loads(f.read())
    dataset = dataset['applications']

print(dataset)


# 데이프레임화
application_df = pd.DataFrame(dataset)
application_df.head()

In [ ]:
application_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61 entries, 0 to 60
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id            61 non-null     object 
 1   candidateId    61 non-null     object 
 2   pnp            61 non-null     object 
 3   applyMajor1    61 non-null     object 
 4   applyMajor2    61 non-null     object 
 5   applySemester  61 non-null     object 
 6   applyTimes     61 non-null     object 
 7   applyGPA       61 non-null     float64
 8   applyGrade     61 non-null     object 
 9   __v            61 non-null     int64  
dtypes: float64(1), int64(1), object(8)
memory usage: 4.9+ KB


### 1.2 데이터프레임 중 의미 없는 Column 제거

[ 데이터셋 중 특수한 Column ]

(1) **user_df**
- name: 한국어
- firstMajor: 한국어, object 데이터 타입 (딕셔너리)
- secondMajor: 한국어, object 데이터 타입 (딕셔너리)
- email: 특수기호 포함
- nickname: (not all) 한국어
- GPA, Semester: 특수기호 포함 (합격자/지원자 여부에 따라 상이한 column)


(2) **application_df**
- applyMajor1: 한국어, object 데이터 타입 (딕셔너리)
- applyMajor2: 한국어, object 데이터 타입 (딕셔너리)
- applySemester: 특수기호 - 포함
- applyGrade: 특수기호 - 포함
- applyGPA: 특수기호 . 포함

In [ ]:
# application_df 전처리 (현재 총 61개의 샘플)

drop_list = ['pnp', '__v', '_id']
application_df.drop(labels=drop_list, axis=1, inplace=True)

print(application_df.info())
# print(application_df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61 entries, 0 to 60
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   candidateId    61 non-null     object 
 1   applyMajor1    61 non-null     object 
 2   applyMajor2    61 non-null     object 
 3   applySemester  61 non-null     object 
 4   applyTimes     61 non-null     object 
 5   applyGPA       61 non-null     float64
 6   applyGrade     61 non-null     object 
dtypes: float64(1), object(6)
memory usage: 3.5+ KB
None


In [ ]:
# application_df 의 'candidateId' column name 변경

application_df.rename(columns = {"candidateId": "_id"}, inplace=True)
print(application_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61 entries, 0 to 60
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id            61 non-null     object 
 1   applyMajor1    61 non-null     object 
 2   applyMajor2    61 non-null     object 
 3   applySemester  61 non-null     object 
 4   applyTimes     61 non-null     object 
 5   applyGPA       61 non-null     float64
 6   applyGrade     61 non-null     object 
dtypes: float64(1), object(6)
memory usage: 3.5+ KB
None


In [ ]:
# user_df 전처리 (현재 총 242개의 샘플)

# user_df 에서 공통으로 제거해야하는 column 제거
drop_list = ['__v', 'name', 'email', 'nickname', 'profilePic', 'leave', 'studentId', 'totalReport', 'profileName']
user_df.drop(labels=drop_list, axis=1, inplace=True)

# user_df 의 경우, 회원구분에 따라 세부 column 이 다르기 때문에, 해당 데이터 분리
user_passer_df = user_df[user_df['role'] == 'passer']
user_candidate_df = user_df[user_df['role'] == 'candidate']

print(user_passer_df.info())
print(user_candidate_df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19 entries, 0 to 236
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   _id           19 non-null     object 
 1   firstMajor    19 non-null     object 
 2   role          19 non-null     object 
 3   secondMajor   19 non-null     object 
 4   passSemester  19 non-null     object 
 5   passGPA       19 non-null     float64
 6   hopeMajor1    0 non-null      object 
 7   hopeMajor2    0 non-null      object 
 8   hopeSemester  0 non-null      object 
 9   curGPA        0 non-null      float64
 10  changeGPA     0 non-null      float64
 11  isApplied     0 non-null      object 
dtypes: float64(3), object(9)
memory usage: 1.9+ KB
None
<class 'pandas.core.frame.DataFrame'>
Int64Index: 223 entries, 1 to 241
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   _id           223 non-null    object 
 1 

결과적으로, passer 샘플은 총 19개 candidate 샘플은 총 223 개 존재한다. \
passer 샘플은 그 개수가 너무 적으므로, 일단 candidate data 를 중점적으로 활용한다.

In [ ]:
# user_passer_df 의 불필요한 column 정리
drop_list1 = ['role', 'hopeMajor1', 'hopeMajor2', 'hopeSemester', 'curGPA', 'changeGPA', 'isApplied']
user_passer_df.drop(labels=drop_list1, axis=1, inplace=True)

# user_candidate_df 의 불필요한 column 정리
drop_list2 = ['role', 'secondMajor', 'passSemester', 'passGPA', 'changeGPA', 'curGPA', 'hopeSemester', 'isApplied']
user_candidate_df.drop(labels=drop_list2, axis=1, inplace=True)

In [ ]:
print(user_passer_df.info())
print(user_candidate_df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19 entries, 0 to 236
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   _id           19 non-null     object 
 1   firstMajor    19 non-null     object 
 2   secondMajor   19 non-null     object 
 3   passSemester  19 non-null     object 
 4   passGPA       19 non-null     float64
dtypes: float64(1), object(4)
memory usage: 912.0+ bytes
None
<class 'pandas.core.frame.DataFrame'>
Int64Index: 223 entries, 1 to 241
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   _id         223 non-null    object
 1   firstMajor  223 non-null    object
 2   hopeMajor1  223 non-null    object
 3   hopeMajor2  223 non-null    object
dtypes: object(4)
memory usage: 8.7+ KB
None


### 1.3 결측치 확인 및 제거

In [ ]:
# df 별 결측치 확인
print(user_candidate_df.isnull().sum())
print(application_df.isnull().sum())

_id           0
firstMajor    0
hopeMajor1    0
hopeMajor2    0
dtype: int64
_id              0
applyMajor1      0
applyMajor2      0
applySemester    0
applyTimes       0
applyGPA         0
applyGrade       0
dtype: int64


결측치가 존재하지 않으므로, 이 부분에 대해서는 별도의 처리를 해주지 않는다.

### 1.4 Major column 의 딕셔너리 데이터값 한 가지로 통일

In [ ]:
# name KEY 에 해당하는 value 추출 함수 정의

def extract_name(major_dict):
    return major_dict['name'] if 'name' in major_dict else None

In [ ]:
# application_df 처리
# 처리필요한 열: applyMajor1, applyMajor2
# (_id / name)

# temp = application_df.copy()
application_df['applyMajor1'] = application_df['applyMajor1'].apply(extract_name)
application_df['applyMajor2'] = application_df['applyMajor2'].apply(extract_name)

application_df.head()

In [ ]:
# user_candidate_df 처리
# 처리필요한 열: firstMajor, hopeMajor1, hopeMajor2
# (_id / name)

# temp = user_candidate_df.copy()
user_candidate_df['firstMajor'] = user_candidate_df['firstMajor'].apply(extract_name)
user_candidate_df['hopeMajor1'] = user_candidate_df['hopeMajor1'].apply(extract_name)
user_candidate_df['hopeMajor2'] = user_candidate_df['hopeMajor2'].apply(extract_name)

user_candidate_df.head()

### 1.5 모의지원 데이터 및 회원정보 데이터 병합 (merge)

In [ ]:
merged_df = pd.merge(application_df, user_candidate_df)
print(merged_df.info())
print(merged_df.head())

In [ ]:
merged_df.head()

In [ ]:
# CSV 파일로 저장
merged_df.to_csv('/content/drive/MyDrive/Colab Notebooks/kuply-MLOps/merged_df.csv', index=None)

# 2. Model Fitting

[ 파이프라인 개요 ]

1. 임베딩
2. 차원축소
3. 클러스터링


# 3. Evaluation

In [ ]:
# 데이터 불러오기

df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/kupply-MLOps/merged_df.csv')
df.head()

_id applyMajor1 applyMajor2 applySemester applyTimes  \
0  6546e6d2d99b2247f05a18a0       컴퓨터학과        통계학과        2023-2      First   
1  6544e32ed99b2247f0428e27        심리학부       미디어학부        2023-2      First   
2  65439ec765045ea661f3e461       컴퓨터학과        통계학과        2023-2      First   
3  6543a22565045ea661f465a0        경영학과       희망 없음        2023-2      First   
4  65486c7a1eb02ee3ad22be9c        경제학과       희망 없음        2023-2      First   

   applyGPA applyGrade firstMajor hopeMajor1 hopeMajor2  
0      4.11        3-2    산업경영공학부      컴퓨터학과       통계학과  
1      4.14        2-2    디자인조형학부       심리학부      미디어학부  
2      3.90        3-2        수학과      컴퓨터학과       통계학과  
3      4.32        2-1       경제학과       경영학과      희망 없음  
4      4.37        2-2     국어국문학과       경제학과      희망 없음

In [ ]:
# 라이브러리 임포트

from sklearn import datasets,metrics
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.datasets import load_iris
import numpy as np
from sklearn.metrics import silhouette_score
from scipy.spatial import distance # 거리 계산

from google.colab import files
from IPython.display import Image